### PR TITLE
[OpDevOptimize] Add common infershape functions

### PR DESCRIPTION
**Pull request by zhiqiu**
### PR types

Others
### PR changes

OPs
### Describe

Add common infershape functions

Currently, there are three common infershape functions:

- `UnaryOpUnchangedInferShape`:  shape input(0) -> output(0) without change.
- `UnaryOpUnchangedInferShapeCheckAxis`:  shape input(0) -> output(0) without change, check if axis in range [-Rank(x), Rank(x)-1]
- `BinaryOpBroadcastInferShape`:  broadcast input(0) and input(1) -> output(0)

We are going to add more.

#### Usage
With these common infershape functions, when developing a new operator and overridding the infershape function, we can check if one of the common infershape functions satisfies. If yes, just reuse the common one.

For example, the `selu` op has infershape function as below,

```
  void InferShape(framework::InferShapeContext *ctx) const override {	  
    OP_INOUT_CHECK(ctx->HasInput("X"), "Input", "X", "selu");	   
    OP_INOUT_CHECK(ctx->HasOutput("Out"), "Output", "Out", "selu");	

    ctx->ShareDim("X", /*->*/ "Out");	
    ctx->ShareLoD("X", /*->*/ "Out");	
  }
```
and, it can be refined as,
```
  void InferShape(framework::InferShapeContext *ctx) const override {	  
     return UnaryOpUnchangedInferShape(ctx);
  }
```

#### Advantage

- improve the efficiency of development, reduce the code of developing a new operator,
- improve the quality of code, reduce errors,
- improve the maintainability of the whole system.

### CODE REVIEW

File: paddle/fluid/operators/common_infer_shape_functions.cc [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+  ctx->ShareDim(ctx->GetInputNameByIdx(0), /*->*/ ctx->GetOutputNameByIdx(0));
+  ctx->ShareLoD(ctx->GetInputNameByIdx(0), /*->*/ ctx->GetOutputNameByIdx(0));
+}
+
+// broadcast input(0) and input(1) -> output(0)
+void BinaryOpBroadcastInferShape(framework::InferShapeContext *ctx) {
+  auto x_name = ctx->GetInputNameByIdx(0);
+  auto y_name = ctx->GetInputNameByIdx(1);
+  auto out_name = ctx->GetOutputNameByIdx(0);
+  OP_INOUT_CHECK(ctx->HasInput(x_name), "Input", x_name, "ElementwiseOp");
```
Comment (chenwhql) [inline paddle/fluid/operators/common_infer_shape_functions.cc]: If `OP_INOUT_CHECK` is not suitable here, you can directly use `PADDLE_ENFORCE_**`

File: paddle/fluid/operators/common_infer_shape_functions.cc [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+          "The input var's type should be LoDTensor, but the "
+          "received is %s [%s].",
+          ctx->GetInputsVarType(y_name).front(), ctx->Inputs(y_name).front()));
+
+  if (ctx->GetInputsVarType(x_name).front() ==
+      framework::proto::VarType::SELECTED_ROWS) {
+    PADDLE_ENFORCE_EQ(
+        y_dims.size(), 1u,
+        platform::errors::InvalidArgument(
+            "For elementwise_op, if X is Sparse(VarType.SELECTED_ROWS"
```
Comment (chenwhql) [inline paddle/fluid/operators/common_infer_shape_functions.cc]: only used for `elementwise op`?

File: paddle/fluid/operators/common_infer_shape_functions.cc [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+    int axis = ctx->Attrs().Get<int>("axis");
+    axis = (axis == -1 ? std::abs(x_dims.size() - y_dims.size()) : axis);
+    std::vector<int> x_dims_array(max_dim);
+    std::vector<int> y_dims_array(max_dim);
+    std::vector<int> out_dims_array(max_dim);
+    details::GetBroadcastDimsArrays(x_dims, y_dims, x_dims_array.data(),
+                                    y_dims_array.data(), out_dims_array.data(),
+                                    max_dim, axis);
+    ctx->SetOutputDim(out_name, framework::make_ddim(out_dims_array));
+    // to do
```
Comment (chenwhql) [inline paddle/fluid/operators/common_infer_shape_functions.cc]: to do what? need to write more here?

Comment (zhiqiu) [reply paddle/fluid/operators/common_infer_shape_functions.cc]: Removed, thx.

File: paddle/fluid/operators/common_infer_shape_functions.cc (hunk 2 above)
Comment (zhiqiu) [reply paddle/fluid/operators/common_infer_shape_functions.cc]: Changed the error message, thx.

File: paddle/fluid/operators/common_infer_shape_functions.cc (hunk 1 above)
Comment (zhiqiu) [reply paddle/fluid/operators/common_infer_shape_functions.cc]: Removed as discussed offline, thx.

File: paddle/fluid/operators/common_infer_shape_functions.cc [hunk 4]
```diff
@@ -98,49 +101,46 @@ void UnaryOpUnchangedInferShapeCheckAxis(framework::InferShapeContext *ctx) {
...
 
 // broadcast input(0) and input(1) -> output(0)
 void BinaryOpBroadcastInferShape(framework::InferShapeContext *ctx) {
   auto x_name = ctx->GetInputNameByIdx(0);
   auto y_name = ctx->GetInputNameByIdx(1);
   auto out_name = ctx->GetOutputNameByIdx(0);
-  OP_INOUT_CHECK(ctx->HasInput(x_name), "Input", x_name, "ElementwiseOp");
-  OP_INOUT_CHECK(ctx->HasInput(y_name), "Input", y_name, "ElementwiseOp");
-  OP_INOUT_CHECK(ctx->HasOutput(out_name), "Output", out_name, "ElementwiseOp");
   auto x_dims = ctx->GetInputDim(x_name);
```
Comment (chenwhql) [inline paddle/fluid/operators/common_infer_shape_functions.cc:113]: not important, I think we can add some comment, tell developers "why not need to check HasInput and HasOutput here"

**Review by chenwhql — APPROVED** [review]
LGTM

File: paddle/fluid/operators/common_infer_shape_functions.cc (hunk 4 above)
Comment (zhiqiu) [reply paddle/fluid/operators/common_infer_shape_functions.cc:113]: Thanks, I will add comments in the next PR.

**Pull request merged**